### PR TITLE
Better encapsulate diesel's connection to ease the split of editoast_models

### PR DESCRIPTION
**Pull request by Wadjetz**
Create an abstraction of the diesel database connection and transactions

part of the #6980

How to test this PR:

- [ ] Make sure everything work as before
- [ ] No deadlocks in tests

### CODE REVIEW

File: editoast/editoast_models/src/db_connection_pool.rs [hunk 1]
```diff
@@ -17,21 +21,81 @@ use openssl::ssl::SslMethod;
...
+                let mut handle = self.write().await;
+                match TxManager::rollback_transaction(handle.deref_mut()).await {
+                    Ok(()) | Err(diesel::result::Error::BrokenTransactionManager) => {
+                        Err(callback_error)
+                    }
+                    Err(rollback_error) => Err(rollback_error.into()),
+                }
+            }
+        }
+    }
```
Comment (woshilapin) [inline editoast/editoast_models/src/db_connection_pool.rs:89]: I believe this piece of code takes inspiration from the existing implementation inside `diesel` crate (see [`TransactionManager::transaction()`](https://docs.rs/diesel/2.1.6/src/diesel/connection/transaction_manager.rs.html#50-71)). It's probably a good idea to keep a link to it and refer to it so the next developer that needs to maintain this function know it should also take inspiration on the `diesel` crate. Maybe also add a comment on the reason why we have to reimplement it ourselves and we can't use the existing one.

File: editoast/src/modelsv2/fixtures.rs [hunk 2]
```diff
@@ -193,9 +194,6 @@ pub fn rolling_stock_livery_changeset(
     rolling_stock_id: i64,
     compound_image_id: i64,
 ) -> Changeset<RollingStockLiveryModel> {
-    // let rolling_stock = named_fast_rolling_stock(&rs_name, db_pool.clone()).await;
-    // let image = document_example(db_pool.clone()).await;
```
Comment (woshilapin) [inline editoast/src/modelsv2/fixtures.rs:197]: Can be removed.

File: editoast/src/modelsv2/fixtures.rs [hunk 3]
```diff
@@ -310,3 +308,15 @@ pub async fn create_work_schedule_group(conn: &mut DbConnection) -> WorkSchedule
...
+    include_str!("../tests/train_schedules/simple_array.json")
+}
+
+pub fn get_fast_rolling_stock_schema(name: &str) -> RollingStock {
+    let mut rolling_stock_form: RollingStock =
+        serde_json::from_str(include_str!("../tests/example_rolling_stock_1.json"))
+            .expect("Unable to parse");
+    rolling_stock_form.name = name.to_string();
+    rolling_stock_form
+}
```
Comment (woshilapin) [inline editoast/src/modelsv2/fixtures.rs]: Why do we need these new functions in a PR that only change the `DbConnection`?

**Review by woshilapin — CHANGES_REQUESTED** [review]
Big work, will be much better. Thank you 🎉 

I request a change on the documentation of the re-implementation of `transaction` since it's a central but non-trivial function of this new `DbConnection`. I believe it's important to give as much context as possible to ease future maintenance.

File: editoast/src/modelsv2/fixtures.rs (hunk 3 above)
Comment (woshilapin) [reply editoast/src/modelsv2/fixtures.rs]: Removed in https://github.com/OpenRailAssociation/osrd/pull/8712/commits/e8c0350489cfc4c6543056d5fb81ba53621c7e21.

File: editoast/editoast_models/src/db_connection_pool.rs (hunk 1 above)
Comment (woshilapin) [reply editoast/editoast_models/src/db_connection_pool.rs:89]: Proposition of documentation in https://github.com/OpenRailAssociation/osrd/pull/8712/commits/9fef9a6a13826c3c8cf2c5ba809dc1aa61b2314c. @leovalais What do you think?

File: editoast/editoast_derive/src/model.rs
Comment (leovalais) [inline editoast/editoast_derive/src/model.rs:1]: Unless I misunderstood something, `DieselConnection` is supposed to be an `editoast_models` internal type. I'd say that these functions should take a `DbConnection` just like before instead.

File: editoast/editoast_models/src/lib.rs [hunk 4]
```diff
@@ -1,16 +1,23 @@
...
 /// Useful for functions which only points of failure are the DB calls.
 #[derive(Debug, thiserror::Error)]
-#[error("an error occured while querying the database: {0}")]
-pub struct DatabaseError(#[from] diesel::result::Error);
+pub enum DatabaseError {
+    #[error("an error occured while querying the database: {0}")]
+    DieselError(#[from] diesel::result::Error),
+    #[error("an error occured while retrieving a connection from the pool: {0}")]
+    DatabasePoolError(#[from] DatabasePoolError),
+}
```
Comment (leovalais) [inline editoast/editoast_models/src/lib.rs]: I don't understand this change. There was two different types `DatabaseError` and `DatabasePoolError` that

1. serve different purposes, cf. their error messages
2. allows `editoast_models` to forward errors without exposing diesel's errors

Besides, `DatabasePoolError` also forwards `diesel::result::Error`. Its seems to me that the error types don't require any modification.

File: editoast/editoast_models/src/lib.rs [hunk 5]
```diff
@@ -1,16 +1,23 @@
...
 
 pub mod db_connection_pool;
 pub mod tables;
 
+pub use db_connection_pool::DbConnection;
 pub use db_connection_pool::DbConnectionPoolV2;
 
-pub type DbConnection = AsyncPgConnection;
-pub type DbConnectionPool = Pool<DbConnection>;
+pub type DieselConnection = AsyncPgConnection;
```
Comment (leovalais) [inline editoast/editoast_models/src/lib.rs]: Why is this type still `pub`? We'd like to only expose `DbConnection` outside of the crate. Am I missing something?

File: editoast/src/modelsv2/auth.rs [hunk 6]
```diff
@@ -39,49 +41,50 @@ impl<B: BuiltinRoleSet + Send + Sync> StorageDriver for PgAuthDriver<B> {
...
-                                authn_user::name.eq(&user.name),
-                            ))
-                            .execute(conn)
-                            .await?;
-
-                        Ok(id)
+        self.pool
+            .get()
+            .await?
+            .clone()
```
Comment (leovalais) [inline editoast/src/modelsv2/auth.rs]: `clone`?

**Review by leovalais — CHANGES_REQUESTED** [review]
Huge work! I have a few comments about the API, especially for errors.

That's work for another PR since this one is already massive:

1. Most functions don't require a `conn: &mut DbConnection` but rather just a `conn: DbConnection` which became cheap to clone.
2. Maybe we could implement `Future` for `DbConnection` so that `conn.await` <=> `conn.write().await.deref_mut()`
3. Some sync test util functions might declutter the boilerplate the `RwLock` introduced.

File: editoast/editoast_derive/src/model.rs
Comment (leovalais) [reply editoast/editoast_derive/src/model.rs:1]: Oops scratch that... it's Model V1

**Review by leovalais — APPROVED** [review]
Approving in order not to block the PR, but since I made the modifications myself, this doesn't account for much 🤣

**Review by woshilapin — APPROVED** [review]

**Pull request merged**